### DISCUSSION TITLE
How to set radiobutton values

### DISCUSSION THREAD

**Discussion by fernandesjeevan** [py-pdf/pypdf#2804] [Q&A] [answered]
Hey guys i have to fill pdf using data from json file, so far i have been able to do it for textfield,dropdown boxes, checkboxes as it was straightforward, I have not been able to fill out radio buttons with it. 
[sample_pdf 3.pdf](https://github.com/user-attachments/files/16655498/sample_pdf.3.pdf)

```
field_values = {
    'radio_fields': {
        'Group6': 'Choice1',  
        
    }
}
for i in range(len(page["/Annots"])):
    annotation = page["/Annots"][i].get_object()
    actual_field_name = str(annotation.get('/T'))

    # Update radio button fields
    if actual_field_name in field_values['radio_fields']:
        selected_option = field_values['radio_fields'][actual_field_name]
        annotation.update({
            NameObject('/V'): NameObject(selected_option),
            NameObject('/AS'): NameObject(selected_option)
        })
```
I have tried this, I don't know whether it is correct, and also this,
`hobby = "reading"
if hobby=="reading":
      writer.update_page_form_field_values(writer.pages[0],{'Choice1':"/Yes"})`

here is the radiobutton object of the pdf
`'Group6': {'/T': 'Group6', '/FT': '/Btn', '/Ff': 49152, '/Kids': [IndirectObject(80, 0, 2808974460832), IndirectObject(81, 0, 2808974460832), IndirectObject(82, 0, 2808974460832), IndirectObject(83, 0, 2808974460832)], '/Opt': ['Choice1', 'Choice1', 'Choice1', 'Choice1']}}`

can you please help me out on this, I am very new to this.

**Comment by pubpub-zz** [comment]
first please ensure you are on the latest build
as you want to modify the doc you should directly create a PdfWriter object:
`w =pypdf.PdfWriter("e:/Downloads/sample_pdf.3.pdf")`

if you run `w.get_fields()` you will get:
{'Name': {'/T': 'Name', '/FT': '/Tx', '/TU': 'Please enter your name here', '/Ff': 8388608}, 'Address': {'/T': 'Address', '/FT': '/Tx', '/TU': 'Please enter your address here', '/Ff': 8388608}, 'Dropdown1': {'/T': 'Dropdown1', '/FT': '/Ch', '/TU': 'Please select the day of the month here', '/Ff': 131072, '/V': '1', '/DV': '1', '/Opt': ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'], '/_States_': ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31']}, 'Dropdown2': {'/T': 'Dropdown2', '/FT': '/Ch', '/TU': 'Please select a month here', '/Ff': 131072, '/V': 'Jan', '/DV': 'Jan', '/Opt': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'], '/_States_': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']}, 'Dropdown3': {'/T': 'Dropdown3', '/FT': '/Ch', '/TU': 'Please select a year here', '/Ff': 131072, '/V': '2012', '/DV': '2012', '/Opt': ['2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024', '2025'], '/_States_': ['2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023', '2024', '2025']}, 'Check Box4': {'/T': 'Check Box4', '/FT': '/Btn', '/_States_': ['/Off', '/Yes']}, 'Check Box1': {'/T': 'Check Box1', '/FT': '/Btn', '/_States_': ['/Off', '/Yes']}, 'Check Box3': {'/T': 'Check Box3', '/FT': '/Btn', '/_States_': ['/Off', '/Yes']}, 'Check Box2': {'/T': 'Check Box2', '/FT': '/Btn', '/_States_': ['/Off', '/Yes']}, 'Text5': {'/T': 'Text5', '/FT': '/Tx', '/TU': 'Please enter your favourite activity here', '/Ff': 12582912}, 'Button7': {'/T': 'Button7', '/FT': '/Btn', '/TU': 'Click here to save this document', '/Ff': 65536, '/Kids': [IndirectObject(85, 0, 1608463209168), IndirectObject(86, 0, 1608463209168)]}, 'Text6': {'/T': 'Text6', '/FT': '/Tx', '/TU': 'Please enter your favourite activity here', '/Ff': 12582912}, 'Group6': {'/T': 'Group6', '/FT': '/Btn', '/Ff': 49152, '/Kids': [IndirectObject(80, 0, 1608463209168), IndirectObject(81, 0, 1608463209168), IndirectObject(82, 0, 1608463209168), IndirectObject(83, 0, 1608463209168)], '/Opt': ['Choice1', 'Choice1', 'Choice1', 'Choice1'], '/_States_': ['/0', '/1', '/2', '/3']}}

for group6, your pdf shows all the same names for the diffrent options, you have then to use the values from _States_:
`w.update_page_form_field_values(None, {"Group6":"/1"})`

**Reply by fernandesjeevan** [reply]
thank you so much for answering the question, hope i am not bothering you, i am extremely sorry but when I run the code 
```
w =pypdf.PdfWriter("sample_pdf 3.pdf")
reader = pypdf.PdfReader("sample_pdf 3.pdf
page = reader.pages[0]
w.add_page(page)
w.update_page_form_field_values(w.pages[0], {"Group6":"/1"})
```

```
Traceback (most recent call last):
  File "C:\Users\jeeva\Downloads\fastapitutorial\pypdf2updatefieds.py", line 16, in <module>
    w.add_page(page)
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\pypdf\_writer.py", line 491, in add_page
    return self._add_page(page, list.append, excluded_keys)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\pypdf\_writer.py", line 422, in _add_page
    page[NameObject(PA.PARENT)] = self._pages
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\PyPDF2\generic\_data_structures.py", line 253, in __setitem__
    raise ValueError("key must be PdfObject")
ValueError: key must be PdfObject
```

i get this error, even though Group6 exists

**Reply by stefan6419846** [reply]
There is something strange about your stacktrace - for some reasons it references both *pypdf* and PyPDF2. Could you please uninstall both (maybe using `--force`) and only install *pypdf* afterwards?

**Reply by fernandesjeevan** [reply]
thank you for replying, but the issue still persists, as it throws error on the version 4.3.1 of pypdf while it doesn't on the version 3.8.1, but the version 3.8.1 does nothing to update the radiofields, sorry for being annoying

```
Traceback (most recent call last):
  File "C:\Users\jeeva\Downloads\fastapitutorial\pypdf2updatefieds.py", line 25, in <module>
    writer.update_page_form_field_values(
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\pypdf\_writer.py", line 1024, in update_page_form_field_values
    writer_parent_annot[NameObject(FA.V)] = TextStringObject(value)
                                            ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\pypdf\generic\_base.py", line 522, in __new__
    if value.startswith(("\xfe\xff", "\xff\xfe")):
       ^^^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'startswith'
```

**Reply by pubpub-zz** [reply]
Your code is wrong:
```
w =pypdf.PdfWriter("sample_pdf.3.pdf") 
w.update_page_form_field_values(0, {"Group6":"/1"})
```

**Reply by fernandesjeevan** [reply]
> w =pypdf.PdfWriter("sample_pdf.3.pdf") 
> w.update_page_form_field_values(0, {"Group6":"/1"})

```
  File "C:\Users\jeeva\Downloads\fastapitutorial\pypdf2radio.py", line 3, in <module>
    w.update_page_form_field_values(0, {"Group6":"/1"})
  File "C:\Users\jeeva\Downloads\fastapitutorial\myenv\Lib\site-packages\pypdf\_writer.py", line 827, in update_page_form_field_values
    if PG.ANNOTS not in page:
       ^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'int' is not iterable
```

**Reply by pubpub-zz** [reply]
And what about  
```w.update_page_form_field_values(None, {"Group6":"/1"})```

**Reply by pubpub-zz** [reply, marked as answer]
My Fault (I was missing my computer to perform the test):
```
w =pypdf.PdfWriter("sample_pdf.3.pdf")
w.update_page_form_field_values(w.pages[0],{"Group6":"/1"})
```
else
```
w =pypdf.PdfWriter("sample_pdf.3.pdf")
w.update_page_form_field_values(None,{"Group6":"/1"})
```
